mysql_mimic/connection.py

```diff
@@ -39,50 +54,422 @@ def __init__(
+
+        stmt_id = next(self.prepared_stmt_seq)
+        num_params = len(REGEX_PARAM.findall(sql))
+
+        stmt = PreparedStatement(
+            stmt_id=stmt_id,
+            sql=sql,
+            num_params=num_params,
+        )
+        self.prepared_stmts[stmt_id] = stmt
+
+        for packet in self.stmt_prepare_ok(stmt):
+            await self.stream.write(packet)
+
+    async def handle_stmt_send_long_data(self, data):
+        """
+        https://dev.mysql.com/doc/internals/en/com-stmt-send-long-data.html
+
+        COM_STMT_SEND_LONG_DATA sends the data for a column.
+        """
+        r = io.BytesIO(data)
+        stmt_id = types.read_uint_4(r)
+        param_id = types.read_uint_2(r)
+        stmt = self.get_stmt(stmt_id)
+        if stmt.param_buffers is None:
+            stmt.param_buffers = {}
+        buffer = stmt.param_buffers.setdefault(param_id, bytearray())
+        buffer.extend(r.read())
+
+    async def handle_stmt_execute(self, data):
+        """
+        https://dev.mysql.com/doc/internals/en/com-stmt-execute.html
+
+        COM_STMT_EXECUTE asks the server to execute a prepared statement as identified by stmt-id.
+        """
+        r = io.BytesIO(data)
+        stmt_id = types.read_uint_4(r)
+        stmt = self.get_stmt(stmt_id)
+        use_cursor = self.read_cursor_flags(r)
+        types.read_uint_4(r)  # iteration count
+        sql = self.interpolate_params(r, stmt)
+        stmt.param_buffers = None
+
+        result_set = await self.session.query(sql)
+
+        if result_set is None:
+            await self.stream.write(self.ok())
+            return
+
+        result_set = ensure_result_set(result_set)
+
+        await self.stream.write(types.uint_len(len(result_set.columns)))
+
+        for column in result_set.columns:
+            await self.stream.write(
+                self.column_definition_41(
+                    name=column.name,
+                    column_type=column.type,
+                    character_set=column.character_set,
+                )
+            )
+
+        rows = (self.binary_resultrow(r) for r in result_set.rows)
+
+        if use_cursor:
+            stmt.cursor = rows
+            await self.stream.write(
+                self.ok_or_eof(flags=types.ServerStatus.SERVER_STATUS_CURSOR_EXISTS)
+            )
+        else:
+            if not self.deprecate_eof():
+                self.eof()
+            for row in rows:
+                await self.stream.write(row)
+            await self.stream.write(self.ok_or_eof())
+
+    async def handle_stmt_fetch(self, data):
+        """
+        https://dev.mysql.com/doc/internals/en/com-stmt-fetch.html
+
+        COM_STMT_FETCH fetches rows from an existing resultset after a COM_STMT_EXECUTE.
+        """
+        r = io.BytesIO(data)
+        stmt_id = types.read_uint_4(r)
+        num_rows = types.read_uint_4(r)
+        stmt = self.get_stmt(stmt_id)
+
+        i = 0
```

Comment:
this is a bit weird, side effect, you set i = 0, but then have python override i with scope from the for loop, can we make this a bit more explicit, i assume you did this for the null case?

Reply:
Haha sure